src/github.rs

```diff
@@ -0,0 +1,167 @@
+use failure::Error;
+use reqwest::{self, blocking::Client, blocking::Response};
+use serde::{Deserialize, Serialize};
+
+use crate::Commit;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct GithubCommitElem { commit: GithubCommit, sha: String }
+#[derive(Serialize, Deserialize, Debug)]
+struct GithubCommit { author: GithubAuthor, committer: GithubAuthor, message: String, }
+#[derive(Serialize, Deserialize, Debug)]
+struct GithubAuthor { date: String, email: String, name: String }
+
+type GitDate = chrono::DateTime<chrono::Utc>;
+
+impl GithubCommitElem {
+    fn date(&self) -> Result<GitDate, Error> {
+        Ok(self.commit.committer.date.parse()?)
+    }
+
+    fn git_commit(self) -> Result<Commit, Error> {
+        let date = self.date()?;
+        Ok(Commit {
+            sha: self.sha,
+            date,
+            summary: self.commit.message,
+        })
+    }
+}
+
+fn headers() -> Result<reqwest::header::HeaderMap, Error> {
+    let mut headers = reqwest::header::HeaderMap::new();
+    let user_agent = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));
+    let user_agent = reqwest::header::HeaderValue::from_static(user_agent);
+    headers.insert(reqwest::header::USER_AGENT, user_agent);
+    if let Ok(token) = std::env::var("GITHUB_TOKEN") {
```

Comment:
So long as the reqwest GET request specifies a `https://` URL, are the headers are always sent securely?  This is a naive question.  I don't know the reqwest library or any of its TLS dependencies well.  It looks like it uses multiple platform-specific TLS backends by default.

Reply:
Honestly, I don't know the answer here either. It would be a good thing to double-check.

Reply:
(Resolving this *is* an example of something that is worth blocking this PR on, I think. I definitely want to be careful about leaking `GITHUB_TOKEN` over insecure channels...)

Reply:
Who is the reqwest guru?

Reply:
AFAIK, this should be fine. reqwest to my knowledge does the right thing by default.

One thing to be a bit cautious with is that the debug formatting of the request will include the token, but we're not publishing that I think so hopefully not a problem.

Reply:
👍

Reply:
> One thing to be a bit cautious with is that the debug formatting of the request will include the token

(Yeah, this is a classic example of where having a `fmt::Display` that delegates to `fmt::Debug` is an anti-pattern. Because it *is* reasonable to want to view the `GITHUB_TOKEN` info as part of local debugging; but you also don't to risk it leaking out into a separate `format!("{}", ...)` invocation...)

Reply:
It sounds like this is resolved then, yes?


----

Out of some degree of paranoia, I surveyed the source code for `reqwest` and didn't notice any obviously wrong uses of `{:?}`.

The two main places I. saw `{:?}` used in the `reqwest` source code for things other than strings and URIs were in:

 1. `#[test]` methods (which is obviously fine)
 2. `log::trace` calls (which is ... probably fine...?)

Reply:
Yes, I think this is resolved. I didn't type this up earlier, but to be clear, this *can* go wrong, but realistically only if the user has misconfigured their system pretty badly -- e.g. adding some random certificates to their trusted list. And even then I'm not entirely sure the native-tls stack will pick it up by default; I think it should, since reqwest (or dependencies) don't ship a certificate bundle to my knowledge.

But that seems like a threat vector we do not care about; if you've done that you're in a not great place and likely leaking a github token as a result is the least of your worries :)